### PR TITLE
Convert from&to native mermaid objects, options and copy/download

### PR DESCRIPTION
**Pull request by rpeyron**
Hello,

I have added some more features:

### Ability to configure default properties settings 

From Extra / Configuration menu set defaultMermaidConfig key with mermaid config structure (see mermaid documentation)
image

In a custom draw.io build you can also set the default configuration in PreConfig.js.

### Convert from & to native mermaid objects 

The mermaid plugin is far easier to use than native support, but one nice thing about the native mermaid objects is that they are correctly displayed on every drawio version, even without the plugin (because it is displayed as SVG images). To get the most of the two, from a contextual menu, a menu will allow converting from or to native mermaid objects if the user needs the shape to be displayed even without the plugin.
image

The full settings will be included in native mermaid objects, so they will be rendered consistently (but some difference may occur because of the different mermaid version).

To avoid problems when changing the shape type I have replaced the "this.state.cell.valueChanged"  in mxShapeMermaid.prototype.updateImage  with corresponding instructions in the parse function. It works well on the tests I made, but I hope it was not useful for use cases that I haven't figured out.

### Download and copy to clipboard

Some links are included in the edit window to be able to download image as SVG or PNG, and copy in clipboard as PNG.
image

I hope you will find these features useful.

Best regards
Remi

### CODE REVIEW

**Comment by nopeslide** [comment]
Hey, thanks again for the PR.
These look like nice features.
Will take a look in a few hours.
Do you use the draw.io desktop version or the vscode integration?

**Comment by rpeyron** [comment]
Hello, I use the draw.io desktop version

File: drawio_desktop/src/mermaid-plugin.js [hunk 1]
```diff
@@ -147,23 +233,160 @@ var DialogMermaid = function (editorUi, shape) {
...
+
+  // Build mermaid settings : by least order
+  // - mermaid_plugin_defaults : this plugin defaults
+  // - EditorUi.defaultMermaidConfig : drawio defaults mermaid
+  // - Editor.config.defaultMermaidConfig : drawio config (from PreConfig and local configuration)
+
+  let mermaid_settings = {};
+  mermaid_settings = mergeDeep(mermaid_settings, mermaid_plugin_defaults);
+  mermaid_settings = mergeDeep(mermaid_settings, window.EditorUi.defaultMermaidConfig);
+  mermaid_settings = mergeDeep(mermaid_settings, window.Editor.config.defaultMermaidConfig);
```
Comment (nopeslide) [inline drawio_desktop/src/mermaid-plugin.js:245]: I had problems here with the latest drawio-desktop version.
```suggestion
  mermaid_settings = mergeDeep(mermaid_settings, mermaid_plugin_defaults);
  try {
    mermaid_settings = mergeDeep(mermaid_settings, window.EditorUi.defaultMermaidConfig);
  } catch (e) {
    if (!e instanceof TypeError) {
      throw e;
    }
  }
  try {
    mermaid_settings = mergeDeep(mermaid_settings, window.Editor.config.defaultMermaidConfig);
  } catch (e) {
    if (!e instanceof TypeError) {
      throw e;
    }
  }
```

**Review by nopeslide — CHANGES_REQUESTED** [review]
Sorry it took longer than anticipated.
Regarding your deep diff/merge code, I would prefer external libs.
At least I propose outsourcing these function since you use them in multiple files.
I also had a problem running this plugin against the latest drawio version,
it completely bricked drawio :D
If we access drawio internals, we should at least do exception handling if drawio moves/renames these internals. 

If you see things differently, please say so.

File: drawio_desktop/src/shapes/shapeMermaid.js [hunk 2]
```diff
@@ -1,9 +1,177 @@
...
+        if ((key in source) && (target[key] === source[key])) return diff
+        // If different, add value to the result
+        return {
+          ...diff,
+          [key]: target[key]
+        }
+      }, {})
+    return diff;
+}
+// ---
```
[the remark this answers is not in the document]
Comment (rpeyron) [reply drawio_desktop/src/shapes/shapeMermaid.js:167]: From what I see in the deep-diff documentation, the result is quite different as the one of this function. But it seems there is another npm package that may work like expected : deep-object-diff (even more used and updated that deep-diff).  I will try this one.

**Comment by rpeyron** [comment]
Hello,  I totally agree with what you suggested, I will work on that this weekend.
About the use of draw.io internals, in addition to the two vars you already fixed I will add some extra checks that are missing when using mermaidData attribute value. And I will also try to do better testing, because the error you encountered actually occurs also on the drawio-desktop version I am using with empty configuration field :-(

**Comment by nopeslide** [comment]
I never used the built in mermaid feature (I wrote this plugin before I knew it existed), so I assume my config field was/is also simply empty.
And again, props to you, your patches make this drawio plugin a full fledged mermaid editor :D
I just love it. When I have time I will copy some of these features into my wavedrom plugin.
I only write js out of necessity, so I'm really glad about your contributions :)

File: drawio_desktop/src/shapes/shapeMermaid.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (nopeslide) [reply drawio_desktop/src/shapes/shapeMermaid.js:167]: My bad, to be honest I just named the first package I found since this is such a common problem

**Comment by rpeyron** [comment]
Hello, I have applied the changes discussed above.

**Review by nopeslide — APPROVED** [review]
nice, lgtm

**Pull request merged**